pymc_experimental/inference/smc/sampling.py

```diff
@@ -0,0 +1,423 @@
+from pymc.sampling.jax import get_jaxified_graph
+from pymc.util import RandomState, _get_seeds_per_chain
+
+log = logging.getLogger(__name__)
+
+
+def sample_with_blackjax_smc(
+    n_particles,
+    random_seed: RandomState = None,
+    kernel: str = "HMC",
+    target_ess: float = 0.8,
+    num_mcmc_steps: int = 10,
+    inner_kernel_params: Optional[dict] = None,
+    model=None,
+    iterations_to_diagnose: int = 100,
+):
+    """Samples using BlackJax's implementation of Sequential Monte Carlo.
+       A summary of the algorithm is:
+
+     1. Initialize :math:`\beta` at zero and stage at zero.
+     2. Generate N samples :math:`S_{\beta}` from the prior (because when :math `\beta = 0` the
+        tempered posterior is the prior).
+     3. Increase :math:`\beta` in order to make the effective sample size equal some predefined
+        value (target_ess)
+     4. Compute a set of N importance weights W. The weights are computed as the ratio of the
+        likelihoods of a sample at stage i+1 and stage i.
+     5. Obtain :math:`S_{w}` by re-sampling according to W.
+     6. Run N independent MCMC chains, starting each one from a different sample
+        in :math:`S_{w}`. For that, set the kernel and inner_kernel_params.
+     7. The N chains are run for num_mcmc_steps each.
+     8. Repeat from step 3 until :math:`\beta \\ge 1`.
+     9. The final result is a collection of N samples from the posterior
+
+    Parameters
+    ----------
+    n_particles
+    random_seed
+    kernel
+    target_ess
+    num_mcmc_steps
+    inner_kernel_params:
+    model: PyMC model to sample from
+    iterations_to_diagnose: Number of iterations to generate diagnosis for.
+
+    Returns
+    -------
+    An Arviz Inference data.
+
+    """
+
+    model = modelcontext(model)
+    random_seed = np.random.default_rng(seed=random_seed)
+
+    if inner_kernel_params is None:
+        inner_kernel_params = {}
+
+    log.info(
+        f"Will only diagnose the first {iterations_to_diagnose} SMC iterations,"
+        f"this number can be increased by setting iterations_to_diagnose parameter"
+        f" in sample_with_blackjax_smc"
+    )
+
+    key = jax.random.PRNGKey(_get_seeds_per_chain(random_seed, 1)[0])
+
+    key, initial_particles_key, iterations_key = jax.random.split(key, 3)
+
+    initial_particles = blackjax_particles_from_pymc_population(
+        model, initialize_population(model, n_particles, random_seed)
+    )
+
+    var_map = var_map_from_model(
+        model, model.initial_point(random_seed=random_seed.integers(2**30))
+    )
+
+    posterior_dimensions = sum(var_map[k][1] for k in var_map)
+
+    if kernel == "HMC":
+        mcmc_kernel = blackjax.mcmc.hmc
+        mcmc_parameters = dict(
+            step_size=inner_kernel_params["step_size"],
+            inverse_mass_matrix=jnp.eye(posterior_dimensions),
+            num_integration_steps=inner_kernel_params["integration_steps"],
+        )
+    elif kernel == "NUTS":
```

Comment:
Maybe add also a random walk metropolis so we can compare with the SMC in PyMC easier? 
Also NUTS-SMC is not very well tested IIUC.

Reply:
That's the idea, but it has been difficult to use the algorithm without any tuning on the inner kernel parameters between SMC runs. I am planning on merging the optimization PR on blackjax first and then use it from here. The SMC in Pymc does apply those optimizations.